lib/upsert/merge_function/PG_Connection.rb

```diff
@@ -4,52 +4,22 @@ class Upsert
   class MergeFunction
     # @private
     class PG_Connection < MergeFunction
+      ERROR_CLASS = PG::Error
       include Postgresql
 
-      def execute(row)
-        first_try = true
-        values = []
-        values += row.selector.values
-        values += row.setter.values
-        hstore_delete_handlers.each do |hstore_delete_handler|
-          values << row.hstore_delete_keys.fetch(hstore_delete_handler.name, [])
-        end
-        Upsert.logger.debug do
-          %{[upsert]\n\tSelector: #{row.selector.inspect}\n\tSetter: #{row.setter.inspect}}
-        end
-        begin
-          connection.execute sql, values.map { |v| connection.bind_value v }
-        rescue PG::Error => pg_error
-          if pg_error.message =~ /function #{name}.* does not exist/i
-            if first_try
-              Upsert.logger.info %{[upsert] Function #{name.inspect} went missing, trying to recreate}
-              first_try = false
-              create!
-              retry
-            else
-              Upsert.logger.info %{[upsert] Failed to create function #{name.inspect} for some reason}
-              raise pg_error
-            end
-          else
-            raise pg_error
-          end
-        end
+      def execute_parameterized(query, args = [])
+        controller.connection.execute(query, args)
       end
 
-      # strangely ? can't be used as a placeholder
-      def sql
-        @sql ||= begin
-          bind_params = []
-          i = 1
-          (selector_keys.length + setter_keys.length).times do
-            bind_params << "$#{i}"
-            i += 1
-          end
-          hstore_delete_handlers.length.times do
-            bind_params << "$#{i}::text[]"
-            i += 1
+      def unique_index_on_selector?
+        return @unique_index_on_selector if defined?(@unique_index_on_selector)
+        @unique_index_on_selector = begin
```

Comment:
incidentally you don't need the `begin`, you can just set `@unique_index_on_selector` on the last line